[15.0][ADD] fieldservice_fsm_person_salary_extended + fieldservice_person_salary_sale_reference

LucPinheiro opened the pull request:
@dalonsod 
@ChristianSantamaria

dalonsod commented:
I'll made my own review once @ChristianSantamaria has finished his one

ChristianSantamaria commented on fsm_order_request_workers/README.rst:
```diff
@@ -0,0 +1,58 @@
+=========================
+FSM Order Request Workers
```
About the name of the addon I think it should start with fieldservice like all the others of the OCA and more descriptive, like fieldservice_fsm_person_salary_extended Idk

ChristianSantamaria replied:
When you finish the improvements redo the readme

ChristianSantamaria commented on fsm_order_request_workers/__manifest__.py:
```diff
@@ -0,0 +1,19 @@
+    "license": "LGPL-3",
+    "version": "15.0.1.0.0",
+    "category": "Field Service",
+    "website": "https://github.com/solvosci/slv-hr",
```
Not slv-hr, should be slv-field-service

ChristianSantamaria commented on fsm_order_request_workers/models/fsm_person.py:
```diff
@@ -0,0 +1,25 @@
+
+    @api.depends('payments_salary','payments_km','payments_diets','payments_advance')
+    def compute_net_salary (self):
+        for net in self:
```
The name of net is incorrect, here you are not traversing nets when the fsm.person is being, you are traversing person, there I would put either record or person

ChristianSantamaria commented on fsm_order_request_workers/oca_dependencies.txt:
```diff
@@ -0,0 +1,3 @@
+helpdesk
+
+helpdesk: https://github.com/OCA/field-service/
```
The helpdesk here is not necessary, only the field-service dependency
And this is not the place for the file oca_dependencies must be at the same height as the addon not inside it

dalonsod replied:
Actually this file should simply be removed. It's already included in previous commit

ChristianSantamaria commented on fsm_order_request_workers/readme/CONTRIBUTORS.rst:
```diff
@@ -0,0 +1,2 @@
+* David Alonso <david.alonso@solvos.es>
```
David should be here?

dalonsod replied:
Yes, please remove me 😁

ChristianSantamaria commented on fsm_order_request_workers/views/fsm_order_request_workers_views.xml:
```diff
@@ -0,0 +1,20 @@
+        <field name="arch" type="xml">
+            <xpath expr="//group//field[@name='person_ids']//tree" position="inside">
+                <field name="vehicle" />
+                <field name="payments_salary" widget="monetary" string='Salary'/>
```
When adding the string attribute in the model itself, it is not necessary to add it again in a view, that is, you could remove it here

ChristianSantamaria replied:
And about the monetary widget I think it is not necessary to add it if the field itself is monetary I am not sure try removing them to see if it behaves the same

ChristianSantamaria commented on fsm_order_request_workers/views/fsm_person_request_workers_views.xml:
```diff
@@ -0,0 +1,20 @@
+        <field name="arch" type="xml">
+            <xpath expr="//tree//field[@name='email']" position="after">
+                <field name="vehicle" />
+                <field name="payments_salary" widget="monetary" string='Salary'/>
```
The same here

ChristianSantamaria commented on fsm_order_request_workers/views/fsm_person_request_workers_views.xml:
```diff
@@ -0,0 +1,20 @@
+            </xpath>
+        </field>
+    </record>
+
```
In addition, it is also necessary to add these fields in the fsm.person form view itself to be able to edit it since it is not possible to edit it from the tree

ChristianSantamaria commented on fsm_order_request_workers/models/fsm_person.py:
```diff
@@ -0,0 +1,25 @@
+    payments_diets = fields.Monetary(currency_field='diets_currency_id', String='Diets')
+    payments_advance = fields.Monetary(currency_field='advance_currency_id', String='Advance')
+    advance_currency_id = fields.Many2one('res.currency')
+    net_salary = fields.Monetary(currency_field='net_salary_currency_id', compute='compute_net_salary', String='Net Salary')
```
The name of the compute functions always begins with _, that is, it should be _compute_net_salary

dalonsod replied:
And I miss `store=True`, to be coherent with later `@api.depends` usage. Also, this field is shown in `tree` view, so stored is a good option (or make it non stored, but computed with a `read_group` call)

ChristianSantamaria commented on fsm_order_request_workers/__manifest__.py:
```diff
@@ -0,0 +1,19 @@
+{
+    "name": "FSM Order Request Workers",
+    "summary": """
+        Added fields in the Request Workers list
```
Try to put a more descriptive summary
And the same in the DESCRIPTION.rst

ChristianSantamaria commented on fsm_order_request_workers/views/fsm_person_request_workers_views.xml:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <record id="fsm_person_request_workers_form" model="ir.ui.view">
```
Just by convention always add view_ in front of each id like view_fsm_person...

ChristianSantamaria commented on fsm_order_request_workers/views/fsm_order_request_workers_views.xml:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <record id="fsm_order_request_workers_form" model="ir.ui.view">
```
Just by convention always add view_ in front of each id like view_fsm_order...

ChristianSantamaria reviewed (changes requested):
@LucPinheiro Some improvements:

PD: @dalonsod I hope I have not left anything 😇

dalonsod commented on fsm_order_request_workers/models/fsm_person.py:
```diff
@@ -0,0 +1,25 @@
+    _inherit = "fsm.person"
+    
+    vehicle = fields.Boolean()
+    salary_currency_id = fields.Many2one('res.currency')
```
It's supposed to be the same currency for every field, because `net_salary` will be calculated without any currency conversion (we'll directly add or subtract amounts, so they will be in the same currency).
So we should define only one currency for all the `Monetary` fields in this model.
Also I also miss the initialization for this field (please see https://github.com/odoo/odoo/blob/15.0/addons/purchase/models/purchase.py#L97)

dalonsod commented:
Locally tested and code review,

- `fsm_order_request_workers` folder must be removed.
- Requirements for the main functionality are not achieved. Salary and vehicle data we need should be different for each field service. And added data is per-contact data.

This last point needs a new analysis, because `person_ids` field can't be actually used with current definition.

ChristianSantamaria commented on fieldservice_fsm_person_salary_extended/security/ir.model.access.csv:
```diff
@@ -0,0 +1,3 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_fsm_salary_user,fsm.salary.user,model_fsm_salary,fieldservice.group_fsm_user,1,1,0,0
```
Is it okay for a user to be able to create record? I think it shouldn't

dalonsod replied:
A FSM User cannot create a FSM order, but can modify it, @ChristianSantamaria could you verify if a FSM User has "Edit" button enabled? In that case, you're right, modifying a FSM order implies working link addition => creation permission for this new model

ChristianSantamaria replied:
You are right, it cannot create but it does have the edit button

ChristianSantamaria commented on fieldservice_fsm_person_salary_extended/views/fsm_order_views.xml:
```diff
@@ -1,20 +1,34 @@
-                <field name="payments_advance" />
-                <field name="net_salary" />
+            <xpath expr="//group[@name='request_workers']" position="attributes">
+                <attribute name="invisible">1</attribute>
```
In these cases, it is better to hide the field than the entire group, I understand that you have made a group with another name but I would keep it inside

ChristianSantamaria replied:
This is still missing, since the option of the new "group" with the name of salary lines is not quite correct

ChristianSantamaria commented on fieldservice_fsm_person_salary_extended/__manifest__.py:
```diff
@@ -12,8 +12,10 @@
+        "security/ir.model.access.csv",
+        "views/fsm_order_views.xml",
+        "views/fsm_salary_views.xml",
+        "views/menu.xml",
```
Rename the menu file specifying the model in the name

ChristianSantamaria commented on fieldservice_fsm_person_salary_extended/i18n/es.po:
```diff
@@ -18,42 +18,114 @@ msgstr ""
-#: model:ir.model.fields,field_description:fieldservice_fsm_person_salary_extended.field_fsm_person__payments_km
+#: model:ir.model.fields,field_description:fieldservice_fsm_person_salary_extended.field_fsm_salary__fsm_order_id
+msgid "Fsm Order"
+msgstr "Fsm Orden de Pedidos"
```
The "fsm" should not be shown

ChristianSantamaria reviewed (changes requested):
@LucPinheiro @dalonsod 
Code review only, not tested.
The commit name is wrong this is not crm
IMPORTANT: in this last commit the fieldservice_sale_reference addon was added, be careful with that, remove that addon so that the commit remains a single addon

ChristianSantamaria commented on fieldservice_fsm_person_salary_extended/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+{
+    "name": "FieldService FSM Person Salary Extended",
+    "summary": """
+        Adds salary fields in fsm.person model
```
Should be change "Summary"

ChristianSantamaria commented on fieldservice_fsm_person_salary_extended/models/fsm_person.py:
```diff
@@ -0,0 +1,10 @@
+    _inherit = "fsm.person"
+    
+    fsm_salary_ids = fields.One2many('fsm.order.salary', 'fsm_person_id')
+  
```
Missing EOF line

ChristianSantamaria commented on fieldservice_fsm_person_salary_extended/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1 @@
+Adds salary fields in fsm.person model.
```
Change with new description

ChristianSantamaria commented on fieldservice_fsm_person_salary_extended/README.rst:
```diff
@@ -0,0 +1,57 @@
+=======================================
```
Regenerate README

ChristianSantamaria reviewed (changes requested):
Locally tested and work correctly, here some improves:

Not reviwed security permissions. @dalonsod will take a look..

ChristianSantamaria commented on fieldservice_fsm_person_salary_extended/views/fsm_order_salary_views.xml:
```diff
@@ -0,0 +1,63 @@
+                <field name="fsm_person_id" />
+                <field name="partner_id" />
+                <field name="vehicle" />
+                <field name="payments_salary" sum="Total" />
```
Change the string "Total" in all fields as it is confusing

ChristianSantamaria commented on fieldservice_fsm_person_salary_extended/views/fsm_order_salary_views.xml:
```diff
@@ -0,0 +1,63 @@
+                <field name="payments_diets" sum="Total" />
+                <field name="payments_advance" sum="Total" />
+                <field name="net_salary" sum="Total" />
+                <field name="currency_id" invisible= '1' sum="Total" />
```
I think that not is necessary this SUM in currency

ChristianSantamaria commented on fieldservice_fsm_person_salary_extended/views/fsm_order_salary_views.xml:
```diff
@@ -0,0 +1,63 @@
+                    <group expand="0" string="Group By">
+                    <filter
+                        name="fsm_order_id"
+                        domain=""
```
Remove the domains emptys

ChristianSantamaria commented on fieldservice_fsm_person_salary_extended/README.rst:
```diff
@@ -0,0 +1,57 @@
+
+|badge1| |badge2| |badge3| 
+
+Adds salary fields in fsm.order.salary model.
```
Change the summary to something simpler, like Adds new model fsm.order.salary that relates to orders and persons
And regenerate

ChristianSantamaria commented on fieldservice_fsm_person_salary_extended/views/fsm_order_salary_views.xml:
```diff
@@ -0,0 +1,63 @@
+                <field name="location_id"/>
+                <field name="scheduled_date_start" />
+                <separator/>
+                <filter string="Fsm Person" name="fsm_person_id" />
```
I dont understund this filters, can there be wages without persons or locations?

ChristianSantamaria commented on fieldservice_fsm_person_salary_extended/views/fsm_order_salary_views.xml:
```diff
@@ -0,0 +1,63 @@
+        <field name="arch" type="xml">
+            <search string="Order Salary">
+                <filter string="My Orders Salary" 
+                    domain="[('fsm_person_id.partner_id.user_ids', '=', uid)]" name="my_orders_salary"/>
```
How does this work?

ChristianSantamaria reviewed (changes requested):
Code review, localy tested.
Adds some improvemets:

Try to find a solution to this, when entering the form view, this is text
![imagen](https://user-images.githubusercontent.com/22427989/230844428-1bef99a6-a391-4350-8ca3-9b2e1b811c68.png)

dalonsod commented on fieldservice_fsm_person_salary_extended/models/fsm_person.py:
```diff
@@ -0,0 +1,9 @@
+class FSMPerson(models.Model):
+    _inherit = "fsm.person"
+    
+    fsm_salary_ids = fields.One2many('fsm.order.salary', 'fsm_person_id')
```
This field is never used, so I'd prefer removing it (actually the whole file)

dalonsod commented on fieldservice_fsm_person_salary_extended/views/fsm_order_salary_menu.xml:
```diff
@@ -0,0 +1,23 @@
+        <field name="context">{
+            'hide_fsm_order_id': False,
+            'hide_location_id': False,
+            'hide_sale_reference_salary': False,
```
I'd replace these three parameters in a single one e.g. `hide_fsm_order_fields`, used for every field

dalonsod commented on fieldservice_fsm_person_salary_extended/views/fsm_order_salary_views.xml:
```diff
@@ -0,0 +1,114 @@
+                    widget="many2onebutton"
+                    invisible="context.get('hide_fsm_order_id', True)"
+                />
+                <field name="sequence" widget="handle" />
```
This field should be hidden in salary report view

dalonsod commented on fieldservice_fsm_person_salary_extended/views/fsm_order_salary_views.xml:
```diff
@@ -0,0 +1,114 @@
+                    groups="fieldservice.group_fsm_dispatcher"
+                    sum="Total" 
+                    />
+                <field name="currency_id" invisible= '1' />
```
Please add `scheduled_date_start` header field to be shown in salary report view

dalonsod commented on fieldservice_fsm_person_salary_extended/models/fsm_order_salary.py:
```diff
@@ -0,0 +1,42 @@
+from odoo import models, fields, api
+
+
+class FSMOrderSalary(models.Model):
```
This field has a strange indentation, different than expected, as other files. Could you check?

dalonsod reviewed (changes requested):
See comments

LucPinheiro committed:
[ADD] fieldservice_person_salary

Replaces workers link to FSM Orders to an extended record with persons
and salary

dalonsod committed:
[ADD] fieldservice_person_salary_sale_reference

Adds sale reference to Person Salary for FSM orders.

dalonsod commented:
Several minor cosmetic changes and refactoring and module renaming, now development splitted in two separated modules:

- Toggle widget for vehicle field.
- Filters for vehicles.
- `fieldservice_person_salary` module rebranding, without sale_reference dependecy.
- New fieldservice_person_salary_sale_reference addon.

@ChristianSantamaria @LucPinheiro could you test? You should uninstall old module, then make `git fetch` and `reset --hard`, so new addons will be available to be tested and installed.

ChristianSantamaria commented on fieldservice_person_salary_sale_reference/models/fsm_order_salary.py:
```diff
@@ -0,0 +1,12 @@
+    _inherit = "fsm.order.salary"
+
+    sale_reference_salary = fields.Char(
+        related="fsm_order_id.sale_id.client_order_ref", 
```
This addon fieldservice_sale_reference adds related reference field then we can do this, much simpler
https://github.com/solvosci/slv-field-service/blob/8bd7bd9f7cf92ae3780026c2557a0c0ed23e59d9/fieldservice_sale_reference/models/fsm_order.py#L10
```suggestion
        related="fsm_order_id.sale_reference", 
```

dalonsod replied:
Comment attended

ChristianSantamaria reviewed (changes requested):
LGTM 👍 
I would only add an improvement:

dalonsod reviewed (approved):
It's time to merge
